nxdrive/osi/extension.py

```diff
@@ -0,0 +1,148 @@
+
+if TYPE_CHECKING:
+    from ..manager import Manager  # noqa
+
+log = getLogger(__name__)
+
+
+class Status(Enum):
+    SYNCED = 1
+    SYNCING = 2
+    CONFLICTED = 3
+    ERROR = 4
+    LOCKED = 5
+    UNSYNCED = 6
+
+
+class ExtensionListener(QTcpServer):
+    """
+    Server listening to the OS extensions.
+
+    This TCP server is instantiated during the Manager.__init__(),
+    and starts listening once the signal Manager.started() is emitted.
+
+    It handles requests coming from any FinderSync extension or overlay DLL instance.
+    These requests are JSON-formatted and follow this pattern:
+    {
+        "command": "<command>",
+        "value": "<value>",  # parameters
+        ...
+    }
+    It will look for the callable associated with the command in its `handlers` dict.
+    """
+
+    listening = pyqtSignal()
+    explorer_name = ""
+
+    def __init__(self, manager: "Manager", *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self.manager = manager
+        self.host = "localhost"
+        self.port = 10650
+        self.handlers: Dict[str, Callable] = {}
+        self.newConnection.connect(self._handle_connection)
+
+    def _listen(self):
+        """
+        Called once the Manager.started() is emitted.
+
+        Starts listening and emits a signal so that the extension can be started.
+        """
+        self.listen(QHostAddress(self.host), self.port)
+        log.debug(f"Listening to {self.explorer_name} on {self.host}:{self.port}")
+        self.listening.emit()
+
+    def _handle_connection(self) -> None:
+        """ Called when an Explorer instance is connecting. """
+        con: QTcpSocket = self.nextPendingConnection()
+
+        if not con or not con.waitForConnected():
+            log.error(
+                f"Unable to open extension handler server socket: {con.errorString()}"
+            )
+            return
+
+        if con.waitForReadyRead():
+            payload = con.readLine()
+
+            try:
+                content = self._parse_payload(payload.data())
+            except:
+                log.debug(f"Unable to decode payload: {payload}")
+            else:
+                response = self._handle_content(content)
+                if response:
+                    con.write(self._format_response(response))
+
+        con.disconnectFromHost()
+        con.waitForDisconnected()
+        del con
+
+    def _parse_payload(self, payload: bytes) -> str:
+        """ Called on the bytes received through the socket. """
+        return force_decode(payload)
+
+    def _format_response(self, response: str) -> bytes:
+        """ Called on the string to send through the socket. """
+        return force_encode(response)
+
+    def _handle_content(self, content: str) -> Optional[str]:
+        """ Called on the parsed payload, runs the handler associated with the command. """
+        try:
+            data = json.loads(content)
+        except:
```

Comment:
Prefer `except Exception`. We will have some clean-up about those "too global" catches :)